save-backend/src/main/kotlin/com/saveourtool/save/backend/service/OrganizationService.kt

```diff
@@ -40,28 +40,73 @@ class OrganizationService(
     }
 
     /**
-     * Mark organization with [organizationName] as deleted
+     * Mark organization with [organization] as deleted
      *
-     * @param organizationName an [Organization]'s name to delete
+     * @param newStatus is new status for [organization]
+     * @param organization is organization in which the status will be changed
      * @return deleted organization
      */
     @Suppress("UnsafeCallOnNullableType")
-    fun deleteOrganization(organizationName: String): Organization = getByName(organizationName)
+    fun changeOrganizationStatus(organization: Organization, newStatus: OrganizationStatus): Organization = organization
         .apply {
-            status = OrganizationStatus.DELETED
+            status = newStatus
         }
         .let {
             organizationRepository.save(it)
         }
 
+    /**
+     * Mark organization [organization] as deleted
+     *
+     * @param organization an [Organization] to delete
+     * @return deleted organization
+     */
+    fun deleteOrganization(organization: Organization): Organization = if (!hasProjects(organization.name)) {
+        changeOrganizationStatus(organization, OrganizationStatus.DELETED)
+    } else {
+        organization
+    }
+
+    /**
+     * Mark organization with [organization] as created.
+     * If an organization was previously banned, then all its projects become deleted.
+     *
+     * @param organization an [Organization] to create
+     * @param organization
+     * @return deleted organization
+     */
+    fun recoverOrganization(organization: Organization): Organization {
+        if (organization.status == OrganizationStatus.BANNED) {
+            projectService.getAllByOrganizationName(organization.name).forEach {
+                it.status = DELETED
+                projectService.updateProject(it)
+            }
+        }
+        return changeOrganizationStatus(organization, OrganizationStatus.CREATED)
+    }
+
+    /**
+     * Mark organization with [organization] and all its projects as banned.
+     *
+     * @param organization an [Organization] to ban
+     * @return deleted organization
+     */
+    fun banOrganization(organization: Organization): Organization {
+        projectService.getAllByOrganizationName(organization.name).forEach {
+            it.status = BANNED
+            projectService.updateProject(it)
+        }
+        return changeOrganizationStatus(organization, OrganizationStatus.BANNED)
+    }
+
     /**
      * @param organizationName the unique name of the organization.
      * @return `true` if this organization has at least one non-deleted project,
      *   `false` otherwise.
      */
     fun hasProjects(organizationName: String): Boolean =
             projectService.getAllByOrganizationName(organizationName).any { project ->
-                project.status != DELETED
+                project.status == CREATED
```

Comment:
`hasProjects` doesn't count BANNED projects, it means that it's allowed to remove organization with banned projects -- is it requirement?

Reply:
yes

Reply:
Only active projects are taken into account here